### PR TITLE
On outer attr not found, avoid knock-down errors

### LINKED ISSUE #118455: A nonexistent inner attribute in `lib.rs` or `main.rs` breaks every procedural macro used within the entire crate
### Code

```Rust
#![some_nonexistent_attribute]

#[derive(Debug)]
pub struct SomeUserCode;
```

### Current output

```Shell
error: cannot find attribute `unsafe_code` in this scope
 --> src/lib.rs:1:4
  |
1 | #![some_nonexistent_attribute]
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: cannot determine resolution for the attribute macro `derive`
 --> src/lib.rs:3:3
  |
3 | #[derive(Debug)]
  |   ^^^^^^
  |
  = note: import resolution is stuck, try simplifying macro imports

error: could not compile `bug-replica` (lib) due to 2 previous errors
```

### Desired output

```Shell
error: cannot find attribute `unsafe_code` in this scope
 --> src/lib.rs:1:4
  |
1 | #![some_nonexistent_attribute]
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `bug-replica` (lib) due to previous error
```

### Rationale and extra context

Having an inner nonexistent attribute in `lib.rs` or `main.rs` shouldn't break every procedural macro in the entire crate.

Usually, one would detect this error just after writing, but when contained within `#[cfg_attr(not(feature = "some-feature", invalid_attribute))]` it can remain undetected for a while due to rust-analyzer not reporting errors for every feature combination, and causing the developer to have to deal with potentially several hundred errors without any clear lead on what caused it.

When this bug is triggered while having and using, for example, a struct deriving `#[derive(Clone, Eq, PartialEq)]` or any other trait, it'll make the compiler emit an error for every attempt to use them. For moderately large projects this means suddenly having hundreds to thousands of errors not related to the actual change from one commit to the next one.

### Other cases

_No response_

### Anything else?

The bug exists at least in versions `1.72.1` (stable) and `1.76.0-nightly (a1a37735c 2023-11-23)`.
Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45f3021f3d7420b1d8e76d3295d5e283

### PR DESCRIPTION
**Pull request by estebank**
When we encounter code like

```rust
#![non_existent]

#[derive(Clone)]
struct S;

fn main() {
    S.clone();
}
```

we avoid emitting errors about the `derive` (which isn't early resolved as normal, but *has* a later resolution) and stop the compiler from advancing to later stages, to avoid knock down errors from the `derive` not being evaluated. Recovering these would be possible, but would produce incorrect errors in the cases where had the outer attribute been present it would have modified the behavior/evaluation of the `derive` attributes.

Furthermore, the fan-out of knock down errors caused by this case is incredibly high, easily causing thousands of errors by the introduction of a single incorrect outer attribute in the crate root.

Fix #118455.

### CODE REVIEW

File: compiler/rustc_resolve/src/macros.rs [hunk 1]
```diff
@@ -627,6 +634,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         &mut self,
         path: &ast::Path,
         kind: Option<MacroKind>,
+        inner_attr: bool,
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/macros.rs:637]: Pls make this an enum with Inner/Outer variants for readability

**Review by TaKO8Ki — COMMENTED** [review]
Other than the suggestion from compiler-errors, r=me

**Comment by chenyukang** [comment]
can we hold on this PR since it will conflict with https://github.com/rust-lang/rust/pull/118533

My PR may suppress more this kind of errors by checking builtin macros and attrs.

**Comment by TaKO8Ki** [comment]
@chenyukang Ok. I will cancel my approval temporarily. @bors r-

**Comment by petrochenkov** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/118533.

**Pull request closed without merging**